### ISSUE TITLE
Search entries emitted too soon

### ISSUE THREAD

**Issue by ifroz** [ldapjs/node-ldapjs#602]
When using the `client.search()` API, I encountered an issue, where my client sometimes missed the first few `searchEntry` events (or even the `end` event).
Apparently, if the LDAP server is much faster than the client, the client does not have enough time to bind event listeners in the callback.

In order to mitigate this issue, I published the fork [ldapjs-with-injectable-emitter](https://github.com/ifroz/node-ldapjs/commit/94716d1e46e86c620b1573a2e77f24cb75254c9e?fbclid=IwAR2eAFBC1x7NPZoJAjuL1HDMN38ruiZ-80EscW8XB-2XL1cwpJZPkcTuFJM),
which, in it's search options, accepts an optional `EventEmitter` instance, so that the user can set up all bindings before actually firing off the request. It can be used as follows:

```javascript
return new Promise((resolve, reject) => {
  const entries = []

  const emitter = new EventEmitter()
  emitter.on('searchEntry', entry => { users.push(entry) })
  emitter.on('error', err => reject(err))
  emitter.on('end', () => resolve(entries))

  client.search('ou=People', { emitter }, (err, res) => {
    if (err) reject(err)

    // Calling res.on(...) here might be too late
    res === emitter // => true
  })
})
```

This fixes my use-case, but the issue still remains for users of `v1.0.2`.
Therefore I'd like to open discussion regarding to either implement this approach in `ldapjs` or come up with another solution, so that `ldapjs` users can mitigate this caveat.

**Comment by jsumners** [comment]
Is this an issue in `ldapjs@next`? Can you replicate it in the unit/integration tests?

**Comment by ifroz** [comment]
- I experience the same issue using `v2.0.0-pre.5` (and `v1.0.2`)
- I have partial success reproducing the issue in unit tests, but [this `setTimeout`-based hanging test](https://github.com/ifroz/node-ldapjs/commit/5239f6c68827f2c25b4589089c199d15bb882412) aims to prove my point.
- In my use-case, I use node `v10.15.0` (I see that some tests on `master` hang in `node >= 10`)

**Comment by jsumners** [comment]
New work should be developed against the `next` branch.

**Comment by ifroz** [comment]
Same behaviour reproduced towards the `next` codebase: [next...ifroz:next](https://github.com/ldapjs/node-ldapjs/compare/next...ifroz:next?expand=1) (at [11868021918](https://github.com/ldapjs/node-ldapjs/commit/11868021918fc5bb0f199399300dfbbc93af9093))

**Comment by rkaw92** [comment]
Hi, I've just authored a patch (linked above) against `next` that addresses this without any API changes, by using a custom EventEmitter that stores unemitted events and only emits them when a listener is added. This is necessary because, while it seems like the original API was meant to emulate that of Node.js Streams, it failed to do so reliably because it lacked a "cork" / "flowing mode" distinction that actual Streams have.

Streams do internal buffering and only start emitting entries when a listener for `data` is added - this is properly called switching into "flowing mode", where new data is emitted as it arrives. This is what we eventually want, but only after ensuring that the programmer has had a chance to register event listeners.

Thanks to the test by @ifroz, I had good assertions ready to verify against.

Please let me know what you think. We'll also get around soon to testing this manually on a specific host where an issue with missing search results (supposedly this) is known to occur predictably.

Also, I've had some issues running the tests locally where 2 tests would fail on `next` right after cloning from the upstream and doing an `npm install`, on Node 12.16.1 - my patch doesn't seem to introduce any additional failures, however, and running the newly-added tests in isolation with `--only` yields success. Any assistance here would be appreciated - shall I paste the test failures I'm having, or can somebody else try the upstream `next` branch and check if they concur?

**Comment by UziTech** [comment]
Great work @rkaw92! If you could create a PR against the `next` branch with that commit the GitHub CI could run the tests and we could see any issues there.

**Comment by rkaw92** [comment]
Update:
* Created a PR, as seen above. Of course the tests pass now in the test environment! :confused: The failures are still reproducible on my machine, but they seem unrelated to the problem or the fix. Pushing these out of scope for now.
* We've found out that the reason for the missing entries on the other machine was, in fact, not this bug. Another bug report + fix can be expected soon...

**Comment by jsumners** [comment]
@rkaw92 I look forward to your next fix. As for the test failing issue, if you can open a new issue with a paste of you test output I'll see if anything jumps out at me.

Closing this issue as solved by #610.

**Comment by jsumners** [comment]
⚠️ This issue has been locked due to age. If you have encountered a recent
problem that seems to be covered by this issue, please open a new issue.

Please include a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)
when opening a new issue.
